dask/base.py

```diff
@@ -132,14 +132,42 @@ def normalize_function(func):
 normalize_token.register(object,
         lambda a: normalize_function(a) if callable(a) else a)
 normalize_token.register(dict, lambda a: tuple(sorted(a.items())))
+
 with ignoring(ImportError):
     import pandas as pd
-    normalize_token.register(pd.DataFrame,
-            lambda a: (id(a), len(a), list(a.columns)))
-    normalize_token.register(pd.Series, lambda a: (id(a), len(a), a.name))
+
+    @partial(normalize_token.register, pd.Index)
+    def normalize_index(ind):
+        return tokenize(ind.name, ind.values)
+
+    @partial(normalize_token.register, pd.Categorical)
+    def normalize_categorical(cat):
+        return tokenize(cat.codes, cat.categories, cat.ordered)
+
+    @partial(normalize_token.register, pd.Series)
+    def normalize_series(s):
+        return tokenize(s.name, s._data.blocks[0].values, s.dtype, s.index)
```

Comment:
Perhaps just `map(normalize_token, ...)` (or better yet, only on the ones known to need normalizing) instead to avoid the str + hash? Not mandatory, just a thought.